File: 37.解数独.cpp
/*
 * @lc app=leetcode.cn id=37 lang=cpp
 *
 * [37] 解数独
 */

#include <vector>
#include <unordered_set>
#include <utility>

class Solution {
public:
    void solveSudoku(std::vector<std::vector<char>>& board) {
        rowRecord.resize(9);
        colRecord.resize(9);
        quadRecord.resize(9);

        for(int r = 0; r < 9; r++){
            for(int c = 0; c < 9; c++){
                const auto& ch = board[r][c];
                if(ch != '.'){
                    rowRecord[r].insert(ch);
                    colRecord[c].insert(ch);
                    quadRecord[r/3*3+c/3].insert(ch);
                }
            }
        }

        DFS(std::forward<std::vector<std::vector<char>>>(board), 0);
    }
private:

    void DFS(std::vector<std::vector<char>>&& board, int idx){
        if(idx >= 81){
            flag = true;
            return;
        }

        int r = idx / 9;
        int c = idx - r*9;
        const auto& ch = board[r][c];

        if(ch != '.'){
            DFS(std::forward<std::vector<std::vector<char>>>(board), idx+1);
            return;
        }
        else{
            auto& rowRec = rowRecord[r];
            auto& colRec = colRecord[c];
            auto& quadRec = quadRecord[r/3*3+c/3];
            for(char select = '1'; select <= '9'; select++){
                if(rowRec.find(select) == rowRec.end()
                    && colRec.find(select) == colRec.end()
                    && quadRec.find(select) == quadRec.end())
                {
                    rowRec.insert(select);
                    colRec.insert(select);
                    quadRec.insert(select);
                    board[r][c] = select;
                    DFS(std::forward<std::vector<std::vector<char>>>(board), idx+1);
                    if(flag){
                        return;
                    }
                    rowRec.erase(select);
                    colRec.erase(select);
                    quadRec.erase(select);
                    board[r][c] = '.';
                }
            }
        }
    }

    std::vector<std::unordered_set<char>> rowRecord;
    std::vector<std::unordered_set<char>> colRecord;
    std::vector<std::unordered_set<char>> quadRecord;
    bool flag = false;
};

File: 37.解数独.rs
/*
 * @lc app=leetcode.cn id=37 lang=rust
 *
 * [37] 解数独
 */

// @lc code=start

use std::collections::HashSet;

fn next_x_y(board: &mut Vec<Vec<char>>, x: usize, y: usize) -> Option<(usize, usize)> {
    for next_y in y..9 {
        for next_x in 0..9 {
            if board[next_y][next_x] == '.' {
                return Some((next_x, next_y));
            }
        }
    }
    None 
}

fn dfs(board: &mut Vec<Vec<char>>, x: usize, y: usize, flag: &mut bool) {
    let mut used_nums = HashSet::new();
    for i in 0..9 {
        if board[y][i] != '.' {
            used_nums.insert(board[y][i]);
        }
        if board[i][x] != '.' {
            used_nums.insert(board[i][x]);
        }
    }
    for i in 0..3 {
        for j in 0..3 {
            if board[y / 3 * 3 + i][x / 3 * 3 + j] != '.' {
                used_nums.insert(board[y / 3 * 3 + i][x / 3 * 3 + j]);
            }
        }
    }

    for num in '1'..='9' {
        if !used_nums.contains(&num) {
            board[y][x] = num;
            if let Some((next_x, next_y)) = next_x_y(board, x, y) {
                dfs(board, next_x, next_y, flag);
                if *flag {
                    return;
                }
            } else {
                *flag = true;
                return;
            }
        }
    }
    board[y][x] = '.';
}

impl Solution {
    pub fn solve_sudoku(board: &mut Vec<Vec<char>>) {
        if let Some((next_x, next_y)) = next_x_y(board, 0, 0) {
            let mut flag = false;
            dfs(board, next_x, next_y, &mut flag);
        }
    }
}
// @lc code=end

